.github/workflows/build-rust-code.yml

```diff
@@ -1,102 +1,148 @@
-          - label: arm32v7
-            make-target: arm32
-            rust-target-path: armv7-unknown-linux-gnueabihf
-          - label: amd64
-            make-target: amd64
-            rust-target-path: x86_64-unknown-linux-gnu
+        component:
+          - label: agent
+            cargo-build-args: --release
+          - label: agent-full
+            cargo-build-args: --release -F agent-full,onvif-feat,opcua-feat,udev-feat
     
     steps:
     - name: Checkout the head commit of the branch
       uses: actions/checkout@v3
       with:
         persist-credentials: false
 
-    - name: Rust install
-      uses: dtolnay/rust-toolchain@master
+    - name: Get version.txt
+      id: version-string
+      run: |
+        echo "version=$(cat version.txt)" >> $GITHUB_OUTPUT
+
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v3
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
+
+    - name: Login to GitHub Container Registry
+      uses: docker/login-action@v3
       with:
-        toolchain: 1.68.1
-        components: clippy, rustfmt
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
 
-    - name: Build production rust for ${{ matrix.arch.label }}
-      env:
-        AGENT_FEATURES: "agent-full onvif-feat opcua-feat udev-feat" 
-      run: |
-        docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
-        make install-cross
-        cross --version
-        make akri-cross-build-${{ matrix.arch.make-target }}
+    - name: Docker meta
+      uses: docker/metadata-action@v5
+      id: meta
+      with:
+        images: ghcr.io/project-akri/akri/${{ matrix.component.label }}
+        labels: |
+          org.opencontainers.image.title=akri-${{matrix.component.label}}
+        tags: |
+          type=ref,event=pr
+          type=semver,pattern={{version}}
+          type=semver,pattern={{major}}.{{minor}}
+          type=raw,value=${{steps.version-string.outputs.version}},enable=${{github.event_name != 'release'}}
 
-    - name: Package build binaries
+    - name: Build and push
+      uses: docker/build-push-action@v5
+      with:
+        context: .
+        push: ${{ github.event_name != 'pull_request' }}
+        build-args: |
+          AKRI_COMPONENT=agent
+          EXTRA_CARGO_ARGS=${{matrix.component.cargo-build-args}}
+        cache-from: type=gha,scope=${{github.ref_name}}-${{matrix.component.label}}
+        cache-to: type=gha,mode=max,scope=${{github.ref_name}}-${{matrix.component.label}}
+        tags: ${{ steps.meta.outputs.tags }}
+        labels: ${{ steps.meta.outputs.labels }}
+        file: build/containers/Dockerfile.rust
+        platforms: linux/amd64,linux/arm64,linux/arm/v7
+
+  build-others:
+    runs-on: ubuntu-latest
+    needs: build-agents
+    timeout-minutes: 50
+    strategy:
+      matrix:
+        component:
+          - label: controller
+          - label: webhook-configuration
```

Comment:
If you look closely at the time frames, you can see that the `agent` builds are taking the entire time, with the other ones only taking a few seconds (as the `cargo build` layer gets pulled from cache).
It's more a 60 minutes build, I put 80 minutes as the limit to give it more headroom in case something get slower on GH side, basically it's the same time as before for the "build Rust", except we had 3x20 minutes so it's now 1x60 minutes as all architectures are built in the same job.

We could get this down by splitting by architecture (as in [this example](https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners)), but this would make a way more complicated workflow file to understand, so I prefer to keep it for a future improvement if it really becomes an issue.

Reply:
That is weird to me that the `agent` container takes so much longer to build. When looking at the build and push step for the other components, they all say `AKRI_COMPONENT=agent`? Are we accidentally only building (and then rebuilding with cache) the agent container?

Reply:
This strategy makes sense. Thanks for walking me through it.

Reply:
That's how it should behave, I noticed an issue after another look after your (deleted it seems) comment about the component arg, will take a look again to ensure it works correctly

Reply:
I fixed the issue, buildkit seems to put a hash of the input parameters in the index cache key, meaning the cache don't get used at all if the parameters are different. Switched to docker local cache and github cache action to make it work as intended.